sylt-compiler/src/typechecker.rs

```diff
@@ -853,6 +885,21 @@ impl TypeChecker {
                                 TypeError::WrongArity { got: args.len(), expected: params.len() }
                             );
                         }
+
+                        if ctx.inside_pure
+                            && !self
+                                .find_node(f)
+                                .constraints
+                                .contains_key(&Constraint::Pure)
```

Comment:
Maybe break this out into a function? We consider `find_node` to be low-level, so it's a good idea to abstract it away. :)